clang/test/Sema/c++-keyword-in-c.c

```diff
@@ -0,0 +1,213 @@
+  int noexcept;         // expected-warning {{identifier 'noexcept' conflicts with a C++ keyword}} \
+                           cxx-error {{expected member name or ';' after declaration specifiers}}
+  int co_await;         // expected-warning {{identifier 'co_await' conflicts with a C++ keyword}} \
+                           cxx-error {{expected member name or ';' after declaration specifiers}}
+  int co_return;        // expected-warning {{identifier 'co_return' conflicts with a C++ keyword}} \
+                           cxx-error {{expected member name or ';' after declaration specifiers}}
+  int co_yield;         // expected-warning {{identifier 'co_yield' conflicts with a C++ keyword}} \
+                           cxx-error {{expected member name or ';' after declaration specifiers}}
+  int consteval;        // expected-warning {{identifier 'consteval' conflicts with a C++ keyword}} \
+                           cxx-error {{consteval can only be used in function declarations}}
+  int constinit;        // expected-warning {{identifier 'constinit' conflicts with a C++ keyword}} \
+                           cxx-error {{constinit can only be used in variable declarations}}
+  int concept;          // expected-warning {{identifier 'concept' conflicts with a C++ keyword}} \
+                           cxx-error {{concept declarations may only appear in global or namespace scope}}
+  int requires;         // expected-warning {{identifier 'requires' conflicts with a C++ keyword}} \
+                           cxx-error {{expected member name or ';' after declaration specifiers}} \
+                           cxx-error {{trailing requires clause can only be used when declaring a function}} \
+                           cxx-error {{expected expression}}
+};
+
+// Smoke test that we catch a keyword used as an enumerator. If we diagnose
+// one, we'll diagnose them all.
+enum E {
+  throw, // expected-warning {{identifier 'throw' conflicts with a C++ keyword}} \
+            cxx-error {{expected identifier}}
+};
+
+// Smoke test that we catch a keyword used as a tag name.
+struct try { // expected-warning {{identifier 'try' conflicts with a C++ keyword}} \
+                cxx-error {{declaration of anonymous struct must be a definition}} \
+                cxx-warning {{declaration does not declare anything}}
+  int x;
+};
+
+// Smoke test that we catch keyword use in a function name.
+void this(void);        // expected-warning {{identifier 'this' conflicts with a C++ keyword}} \
+                           cxx-error {{expected unqualified-id}}
+
+// Smoke test that we catch keyword use in function parameters too.
+void func(int private); // expected-warning {{identifier 'private' conflicts with a C++ keyword}} \
+                           cxx-error {{invalid parameter name: 'private' is a keyword}}
+
+// These are conditionally a keyword in C++, so they're intentionally not being
+// diagnosed as a keyword.
+int module;
```

Comment:
what about `override` and `final`?

Reply:
I'll add them as test cases; they're not intended to be diagnosed either.